File: streaming-bridge.js
/**
 * Simple HTTP Streaming Bridge
 * Receives frames from browser via HTTP POST and streams to YouTube RTMP
 * 
 * Setup:
 * 1. npm install express
 * 2. Set YOUTUBE_STREAM_KEY below
 * 3. Run: node streaming-bridge.js
 * 4. In streaming app: Use endpoint http://localhost:8889/stream
 */

const express = require('express');
const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

const app = express();

// Simple CORS middleware
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.header(
        'Access-Control-Allow-Headers',
        'Content-Type, Authorization, X-Stream-Fps, X-Stream-Width, X-Stream-Height, X-Stream-Id, X-Stream-Label, X-Stream-Frame-Id, X-Stream-Timestamp, X-Stream-Bitrate-Kbps, X-Stream-Maxrate-Kbps, X-Stream-Bufsize-Kbps'
    );
    if (req.method === 'OPTIONS') {
        return res.sendStatus(200);
    }
    next();
});

// ============= CONFIGURATION =============
const PORT = 8889;
const RESTART_DELAY_MS = 3000; // Wait 3 seconds before restarting FFmpeg
const MAX_RESTART_ATTEMPTS = 5; // Max restarts before giving up
const NO_FRAME_TIMEOUT_MS = 12_000; // Restart if no frames for this long

const DEFAULT_STREAM_CONFIG = {
    fps: clampInt(process.env.STREAM_FPS, 5, 60, 30),
    width: clampInt(process.env.STREAM_WIDTH, 320, 3840, 1280),
    height: clampInt(process.env.STREAM_HEIGHT, 240, 2160, 720),
    videoKbps: clampInt(process.env.STREAM_VIDEO_KBPS, 800, 20000, 4500),
    maxKbps: clampInt(process.env.STREAM_MAX_KBPS, 800, 30000, 6000),
    bufKbps: clampInt(process.env.STREAM_BUF_KBPS, 800, 40000, 9000)
};

const STREAM_AUTH = {
    youtubeStreamKey: String(process.env.YOUTUBE_STREAM_KEY || '54c1-px3r-qyp0-sbhv-7bzb'),
    rtmpUrlOverride: String(process.env.YOUTUBE_RTMP_URL || '')
};

const FFMPEG_BIN =
    process.env.FFMPEG_PATH ||
    (fs.existsSync(path.join(__dirname, 'ffmpeg.exe')) ? path.join(__dirname, 'ffmpeg.exe') : 'ffmpeg');
// =========================================

app.use(express.json({ limit: '50mb' }));
app.use(express.raw({ type: 'image/*', limit: '50mb' }));

let ffmpegProcess = null;
let isStreaming = false;
let frameCount = 0;
let restartAttempts = 0;
let restartTimeout = null;
let lastFrameAt = 0;
let currentConfig = { ...DEFAULT_STREAM_CONFIG };
let ffmpegBackpressure = false;

function getYouTubeURL() {
    if (STREAM_AUTH.rtmpUrlOverride) return STREAM_AUTH.rtmpUrlOverride;
    if (!STREAM_AUTH.youtubeStreamKey) return null;
    return `rtmp://a.rtmp.youtube.com/live2/${STREAM_AUTH.youtubeStreamKey}`;
}

function startFFmpeg(config) {
    // Too many restarts - stop trying
    if (restartAttempts >= MAX_RESTART_ATTEMPTS) {
        console.log('[FFmpeg] Too many restart attempts. Stopping. Please check:');
        console.log('[FFmpeg] 1. Is YouTube stream key valid?');
        console.log('[FFmpeg] 2. Is port 1935 (RTMP) open in firewall?');
        console.log('[FFmpeg] 3. Did you enable the stream in YouTube Studio?');
        return false;
    }
    
    const rtmpUrl = getYouTubeURL();
    
    if (!rtmpUrl) {
        console.log('ERROR: YouTube stream key not configured!');
        console.log('Set YOUTUBE_STREAM_KEY environment variable');
        return false;
    }
    
    const cfg = normalizeConfig(config);
    currentConfig = { ...cfg };
    console.log(`[FFmpeg] Starting stream to: ${rtmpUrl} (attempt ${restartAttempts + 1})`);
    console.log(`[FFmpeg] Input: ${cfg.width}x${cfg.height} @ ${cfg.fps} fps, ${cfg.videoKbps} kbps`);
    
    // FFmpeg command to receive JPEG frames and stream to RTMP
    const gop = Math.max(2, Math.floor(cfg.fps * 2));
    const args = [
        '-hide_banner',
        '-loglevel', 'warning',
        '-fflags', '+genpts',
        '-use_wallclock_as_timestamps', '1',
        '-f', 'mjpeg',                 // Input format: MJPEG (JPEG frames)
        '-framerate', String(cfg.fps),
        '-i', '-',                     // Read video from stdin
        '-f', 'lavfi',                 // Add a silent audio source
        '-i', 'anullsrc=channel_layout=stereo:sample_rate=48000',
        '-map', '0:v:0',
        '-map', '1:a:0',
        '-vf', `scale=${cfg.width}:${cfg.height}:flags=bicubic,format=yuv420p`,
        '-c:v', 'libx264',            // H.264 video codec
        '-preset', 'veryfast',        // Fast encoding preset
        '-tune', 'zerolatency',       // Low latency tuning
        '-b:v', `${cfg.videoKbps}k`,  // Video bitrate
        '-maxrate', `${cfg.maxKbps}k`, // Max bitrate
        '-bufsize', `${cfg.bufKbps}k`, // Buffer size
        '-g', String(gop),            // Keyframe interval
        '-keyint_min', String(gop),
        '-r', String(cfg.fps),
        '-c:a', 'aac',                // AAC audio codec
        '-b:a', '128k',              // Audio bitrate
        '-ar', '48000',              // Audio sample rate
        '-ac', '2',
        '-flvflags', 'no_duration_filesize',
        '-f', 'flv',                 // Output format
        rtmpUrl                       // YouTube RTMP URL
    ];
    
    ffmpegProcess = spawn(FFMPEG_BIN, args, { stdio: ['pipe', 'ignore', 'pipe'] });
    
    ffmpegProcess.stderr.on('data', (data) => {
        const msg = data.toString();
        // Only log errors and important messages
        if (msg.includes('error') || msg.includes('Error') || msg.includes('rtmp') || msg.includes('connection') || msg.includes('failed')) {
            console.log('[FFmpeg] ' + msg.trim());
        }
        if (msg.includes('frame=') || msg.includes('speed=')) {
            process.stdout.write('\r[FFmpeg] ' + msg.substring(0, 80));
        }
    });
    
    ffmpegProcess.on('close', (code) => {
        console.log(`\n[FFmpeg] Process exited with code ${code}`);
        ffmpegProcess = null;
        
        // Only auto-restart if we were streaming and lost connection
        if (isStreaming && code !== 0) {
            restartAttempts++;
            console.log(`[FFmpeg] Will retry in ${RESTART_DELAY_MS}ms...`);
            restartTimeout = setTimeout(() => {
                console.log('[FFmpeg] Attempting to restart...');
                startFFmpeg();
            }, RESTART_DELAY_MS);
        } else {
            isStreaming = false;
        }
    });
    
    ffmpegProcess.on('error', (err) => {
        console.log('[FFmpeg] Error:', err.message);
    });
    
    // Handle stdin errors gracefully
    if (ffmpegProcess.stdin) {
        ffmpegProcess.stdin.on('error', (err) => {
            // Ignore EPIPE - happens when FFmpeg closes but we still try to write
            if (err.code !== 'EPIPE') {
                console.log('[FFmpeg stdin] Error:', err.message);
            }
        });
        ffmpegProcess.stdin.on('drain', () => {
            ffmpegBackpressure = false;
        });
    }
    
    return true;
}

function stopFFmpeg(resetAttempts = false) {
    // Clear any pending restart
    if (restartTimeout) {
        clearTimeout(restartTimeout);
        restartTimeout = null;
    }
    if (resetAttempts) restartAttempts = 0;
    
    if (ffmpegProcess) {
        console.log('\n[FFmpeg] Stopping...');
        try {
            if (ffmpegProcess.stdin) {
                ffmpegProcess.stdin.end();
            }
        } catch (e) {}
        try {
            ffmpegProcess.kill();
        } catch (e) {}
        ffmpegProcess = null;
        isStreaming = false;
    }
}

function normalizeConfig(partial) {
    const cfg = { ...DEFAULT_STREAM_CONFIG, ...(partial || {}) };
    cfg.fps = clampInt(cfg.fps, 5, 60, DEFAULT_STREAM_CONFIG.fps);
    cfg.width = clampInt(cfg.width, 320, 3840, DEFAULT_STREAM_CONFIG.width);
    cfg.height = clampInt(cfg.height, 240, 2160, DEFAULT_STREAM_CONFIG.height);
    cfg.videoKbps = clampInt(cfg.videoKbps, 800, 20000, DEFAULT_STREAM_CONFIG.videoKbps);
    cfg.maxKbps = clampInt(cfg.maxKbps, 800, 30000, DEFAULT_STREAM_CONFIG.maxKbps);
    cfg.bufKbps = clampInt(cfg.bufKbps, 800, 40000, DEFAULT_STREAM_CONFIG.bufKbps);
    return cfg;
}

function clampInt(val, min, max, fallback) {
    const n = Number(val);
    if (!Number.isFinite(n)) return fallback;
    return Math.max(min, Math.min(max, Math.floor(n)));
}

function extractConfigFromHeaders(headers) {
    return {
        fps: Number(headers['x-stream-fps'] || headers['x-stream-fps'.toLowerCase()]),
        width: Number(headers['x-stream-width'] || headers['x-stream-width'.toLowerCase()]),
        height: Number(headers['x-stream-height'] || headers['x-stream-height'.toLowerCase()]),
        videoKbps: Number(headers['x-stream-bitrate-kbps'] || headers['x-stream-bitrate-kbps'.toLowerCase()]),
        maxKbps: Number(headers['x-stream-maxrate-kbps'] || headers['x-stream-maxrate-kbps'.toLowerCase()]),
        bufKbps: Number(headers['x-stream-bufsize-kbps'] || headers['x-stream-bufsize-kbps'.toLowerCase()])
    };
}

function configChanged(a, b) {
    return a.fps !== b.fps ||
        a.width !== b.width ||
        a.height !== b.height ||
        a.videoKbps !== b.videoKbps ||
        a.maxKbps !== b.maxKbps ||
        a.bufKbps !== b.bufKbps;
}

// Route: Receive frame from browser
app.post('/stream', (req, res) => {
    // Get frame data first
    let frameData;
    if (req.is('image/*')) {
        frameData = req.body;
    } else if (req.body && req.body.image) {
        // Base64 encoded
        const buffer = Buffer.from(req.body.image, 'base64');
        frameData = buffer;
    }
    
    if (!frameData) {
        return res.status(400).json({ error: 'No image data' });
    }

    const incomingCfg = normalizeConfig(extractConfigFromHeaders(req.headers));
    if (configChanged(currentConfig, incomingCfg) && isStreaming) {
        console.log('[FFmpeg] Config change detected. Restarting FFmpeg...');
        restartAttempts = 0;
        stopFFmpeg();
    }
    
    // Start FFmpeg on first frame if not already running
    if (!isStreaming || !ffmpegProcess) {
        if (!startFFmpeg(incomingCfg)) {
            return res.status(500).json({ error: 'FFmpeg not configured' });
        }
        isStreaming = true;
    }
    
    // Try to write frame
    if (ffmpegProcess && ffmpegProcess.stdin) {
        try {
            if (!ffmpegBackpressure) {
                const ok = ffmpegProcess.stdin.write(frameData);
                if (!ok) ffmpegBackpressure = true;
                frameCount++;
                lastFrameAt = Date.now();
            }
        } catch (e) {
            console.log('[Error] Writing frame:', e.message);
            // Don't auto-restart here - let the close handler deal with it
        }
    }
    
    res.json({ status: 'ok', frames: frameCount });
});

// Route: Get status
app.get('/status', (req, res) => {
    res.json({
        streaming: isStreaming,
        framesReceived: frameCount,
        youtubeConfigured: !!STREAM_AUTH.youtubeStreamKey || !!STREAM_AUTH.rtmpUrlOverride,
        ffmpegRunning: ffmpegProcess !== null,
        restartAttempts: restartAttempts,
        lastFrameAgeMs: lastFrameAt ? Date.now() - lastFrameAt : null,
        currentConfig
    });
});

// Route: Stop stream
app.post('/stop', (req, res) => {
    stopFFmpeg(true);
    frameCount = 0;
    res.json({ status: 'stopped' });
});

// Route: Configure YouTube (alternative to editing file)
app.post('/config', (req, res) => {
    const { streamKey } = req.body;
    if (streamKey) {
        STREAM_AUTH.youtubeStreamKey = String(streamKey);
        console.log('[Config] Stream key received (not persisted)');
        res.json({ status: 'ok' });
    } else {
        res.status(400).json({ error: 'Missing streamKey' });
    }
});

// Main
console.log('=' .repeat(50));
console.log('  Simple HTTP Streaming Bridge');
console.log('=' .repeat(50));
console.log(`Server running on: http://localhost:${PORT}`);
console.log('');
console.log('CONFIGURATION:');
console.log(`- YouTube configured: ${!!STREAM_AUTH.youtubeStreamKey || !!STREAM_AUTH.rtmpUrlOverride}`);
if (!STREAM_AUTH.youtubeStreamKey && !STREAM_AUTH.rtmpUrlOverride) {
    console.log('- WARNING: Set YOUTUBE_STREAM_KEY env var or YOUTUBE_RTMP_URL!');
}
console.log('');
console.log('USAGE:');
console.log(`1. In streaming app, set endpoint: http://localhost:${PORT}/stream`);
console.log('2. Set protocol to: HTTP (FFmpeg Bridge)');
console.log('3. Click Start Stream');
console.log('');
console.log('IMPORTANT:');
console.log('- Make sure to enable your stream in YouTube Studio FIRST');
console.log('- Wait for YouTube to show "Starting" or "Ready" status');
console.log('- Then click Start Stream in this app');
console.log('');
console.log('Press Ctrl+C to stop');
console.log('=' .repeat(50));

app.listen(PORT, () => {
    console.log(`Server ready at http://localhost:${PORT}`);
});

setInterval(() => {
    if (!isStreaming || !ffmpegProcess) return;
    if (lastFrameAt && Date.now() - lastFrameAt > NO_FRAME_TIMEOUT_MS) {
        console.log('[Watchdog] No frames received. Restarting FFmpeg...');
        restartAttempts++;
        stopFFmpeg();
        startFFmpeg(currentConfig);
    }
}, 2000);

process.on('SIGINT', () => {
    console.log('\nShutting down...');
    stopFFmpeg(true);
    process.exit(0);
});

File: streaming-bridge.py
#!/usr/bin/env python3
"""
Local Streaming Bridge - Receives from browser, pushes to YouTube RTMP

Usage:
1. Install Python dependencies: pip install flask opencv-python ffmpeg-python
2. Run: python streaming-bridge.py
3. In app: Use endpoint http://localhost:8889/stream
4. Configure YouTube Stream Key below
"""

import os
import sys
import subprocess
import threading
import time
import io
from flask import Flask, request, Response
import numpy as np
import cv2

# ============= CONFIGURATION =============
YOUTUBE_STREAM_KEY = ""  # Add your YouTube stream key here
# Full RTMP URL: rtmp://a.rtmp.youtube.com/live2/YOUR_KEY
# =========================================

app = Flask(__name__)

# Global frame buffer for FFmpeg
frame_buffer = None
frame_lock = threading.Lock()
stream_active = False
ffmpeg_process = None

def get_youtube_url():
    """Get YouTube RTMP URL with stream key"""
    if not YOUTUBE_STREAM_KEY:
        return None
    return f"rtmp://a.rtmp.youtube.com/live2/{YOUTUBE_STREAM_KEY}"

def start_ffmpeg_stream():
    """Start FFmpeg process to receive frames and stream to YouTube"""
    global ffmpeg_process
    
    youtube_url = get_youtube_url()
    if not youtube_url:
        print("[ERROR] YouTube stream key not configured!")
        print("Please edit streaming-bridge.py and set YOUTUBE_STREAM_KEY")
        return False
    
    print(f"[FFmpeg] Starting stream to: {youtube_url}")
    
    # FFmpeg command - reads from pipe, outputs to RTMP
    ffmpeg_cmd = [
        'ffmpeg',
        '-re',                          # Read at native frame rate
        '-f', 'rawvideo',               # Input format
        '-vcodec', 'rawvideo',          # Raw video codec
        '-s', '1280x720',               # Input resolution
        '-r', '30',                     # Input framerate
        '-pix_fmt', 'bgr24',            # Pixel format
        '-i', '-',                      # Read from stdin
        '-c:v', 'libx264',              # H.264 encoding
        '-preset', 'veryfast',          # Fast encoding
        '-tune', 'zerolatency',         # Low latency
        '-b:v', '4500k',                # Video bitrate
        '-maxrate', '6000k',            # Max bitrate
        '-bufsize', '9000k',            # Buffer size
        '-c:a', 'aac',                  # Audio codec
        '-b:a', '128k',                 # Audio bitrate
        '-ar', '48000',                 # Audio sample rate
        '-f', 'flv',                    # Output format
        youtube_url                     # YouTube RTMP URL
    ]
    
    try:
        ffmpeg_process = subprocess.Popen(
            ffmpeg_cmd,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            bufsize=0
        )
        print("[FFmpeg] Process started successfully")
        return True
    except Exception as e:
        print(f"[ERROR] Failed to start FFmpeg: {e}")
        return False

def stop_ffmpeg_stream():
    """Stop FFmpeg process"""
    global ffmpeg_process
    if ffmpeg_process:
        print("[FFmpeg] Stopping stream...")
        ffmpeg_process.stdin.close()
        ffmpeg_process.terminate()
        ffmpeg_process = None

def send_frame_to_ffmpeg(frame):
    """Send a frame to FFmpeg stdin"""
    global ffmpeg_process
    if ffmpeg_process and ffmpeg_process.stdin:
        try:
            ffmpeg_process.stdin.write(frame.tobytes())
        except Exception as e:
            print(f"[ERROR] Failed to send frame: {e}")
            stop_ffmpeg_stream()

@app.route('/stream', methods=['POST'])
def receive_stream():
    """Receive stream from browser using multipart form data"""
    global stream_active, frame_buffer
    
    if not stream_active:
        # Start FFmpeg when first frame arrives
        if not start_ffmpeg_stream():
            return "Stream not configured", 500
        stream_active = True
    
    # Get frame from request
    if 'frame' not in request.files:
        # Try getting raw image data
        img_data = request.get_data()
        if img_data:
            # Decode image
            nparr = np.frombuffer(img_data, np.uint8)
            frame = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
            if frame is not None:
                send_frame_to_ffmpeg(frame)
    
    return "OK", 200

@app.route('/status', methods=['GET'])
def get_status():
    """Get streaming status"""
    return {
        "active": stream_active,
        "youtube_configured": bool(YOUTUBE_STREAM_KEY),
        "ffmpeg_running": ffmpeg_process is not None
    }

@app.route('/stop', methods=['POST'])
def stop_stream():
    """Stop the stream"""
    global stream_active
    stop_ffmpeg_stream()
    stream_active = False
    return "OK", 200

def main():
    print("=" * 50)
    print("  Local Streaming Bridge")
    print("=" * 50)
    print(f"Stream endpoint: http://localhost:8889/stream")
    print(f"Status endpoint: http://localhost:8889/status")
    print(f"Stop endpoint: http://localhost:8889/stop")
    print("")
    
    if not YOUTUBE_STREAM_KEY:
        print("WARNING: YouTube stream key not configured!")
        print("Edit streaming-bridge.py and set YOUTUBE_STREAM_KEY")
        print("")
    
    print("In the streaming app:")
    print("1. Set Protocol to 'WHIP (WebRTC)'")
    print("2. Use custom endpoint: http://localhost:8889/stream")
    print("3. Click Start Stream")
    print("")
    print("Press Ctrl+C to stop")
    
    # Note: This simple version doesn't actually receive WebRTC
    # For full WebRTC support, you'd need a proper WHIP server
    # This is a placeholder showing the architecture
    
    app.run(host='0.0.0.0', port=8889, debug=False)

if __name__ == '__main__':
    main()

